File: gkr/ch1/1-19.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const maxLine = 1000

func main() {
	rev := make([]byte, maxLine)
	for {
		line, l := getLine(maxLine)
		reverse(line, rev, l)
		fmt.Printf("%s\n", string(rev))
		reset(rev)
	}
}

func reverse(s []byte, rev []byte, lim int) {
	for i, j := lim-1, 0; i >= 0; i, j = i-1, j+1 {
		rev[j] = s[i]
	}
}

func reset(s []byte) {
	for i := range s {
		s[i] = 0
	}
}

func getLine(lim int) ([]byte, int) {
	var c byte
	var i int
	s := make([]byte, lim)
	r := bufio.NewReader(os.Stdin)
	for i = 0; i < lim-1; i++ {
		c, err := r.ReadByte()
		if rune(c) == '\n' || err != nil {
			break
		}
		s[i] = c
	}
	if rune(c) == '\n' {
		s[i] = c
		i++
	}
	return s, i
}

File: kr/ch1/1-19.c
#include <stdio.h>

#define MAXLINE 1000

int getline(char line[], int maxline);
void reverse(char s[], int size);

int main(void)
{
    int len;
    char line[MAXLINE];

    for (int i = 0; i < MAXLINE; i++) {
        line[i] = 0;
    }

    while ((len = getline(line, MAXLINE)) > 0) {
        if (len > 0) {
            reverse(line, len);
            printf("%s", line);
        }
    }

    return 0;
}

void reverse(char s[], int size)
{
    int i, j;

    for (i = 0, j = size - 2; i < j; ++i, --j) {
        char t = s[i];
        s[i]   = s[j];
        s[j]   = t;
    }
}

int getline(char s[], int lim)
{
    int c, i;

    for (i = 0; i < lim - 1 && (c = getchar()) != EOF && c != '\n'; ++i) {
        s[i] = c;
    }

    if (c == '\n') {
        s[i] = c;
        ++i;
    }

    s[i] = '\0';

    return i;
}
